foolbox/attacks/adef_attack.py

```diff
@@ -0,0 +1,398 @@
+        dfdx[c,i,j] = (image[c,i,j+1] - image[c,i,j-1])/2
+        dfdx[c,i,j] = (image[c,i+1,j] - image[c,i-1,j])/2
+
+    positive x-direction is along rows from left to right.
+    positive y-direction is along columns from above to below.
+
+    """
+
+    if color_axis == 2:
+        image = _transpose_image(image)
+    # Derivative in x direction (rows from left to right)
+    dfdx = np.zeros_like(image)
+    # forward difference in first column
+    dfdx[:, :, 0] = image[:, :, 1] - image[:, :, 0]
+    # backwards difference in last column
+    dfdx[:, :, -1] = image[:, :, -1] - image[:, :, -2]
+    # central difference elsewhere
+    dfdx[:, :, 1:-1] = 0.5 * (image[:, :, 2:] - image[:, :, :-2])
+
+    # Derivative in y direction (columns from above to below)
+    dfdy = np.zeros_like(image)
+    # forward difference in first row
+    dfdy[:, 0, :] = image[:, 1, :] - image[:, 0, :]
+    # backwards difference in last row
+    dfdy[:, -1, :] = image[:, -1, :] - image[:, -2, :]
+    # central difference elsewhere
+    dfdy[:, 1:-1, :] = 0.5 * (image[:, 2:, :] - image[:, :-2, :])
+
+    return dfdx, dfdy
+
+
+def _compose(image, vec_field, color_axis):
+    """Calculate the composition of the function image with the vector
+    field vec_field by interpolation.
+
+    new_func = compose(image, vec_field)
+
+    In:
+    image: numpy.ndarray
+        of shape Cxhxw with C=3 or C=1 (color channels), h,w >= 2,
+        and [type] = 'Float' or 'Double'.
+        Contains the values of a function f:R^2 -> R^C
+        on the grid {0,...,h-1}x{0,...,w-1}.
+    vec_field: numpy.array
+        of shape (h,w,2)
+
+    vec_field[y,x,0] is the x-coordinate of the vector vec_field[y,x]
+    vec_field[y,x,1] is the y-coordinate of the vector vec_field[y,x]
+
+    positive x-direction is along rows from left to right
+    positive y-direction is along columns from above to below
+
+    """
+
+    if color_axis == 2:
+        image = _transpose_image(image)
+
+    c, h, w = image.shape  # colors, height, width
+    hrange = np.arange(h)
+    wrange = np.arange(w)
+    MGx, MGy = np.meshgrid(wrange, hrange)
+
+    defMGx = (MGx + vec_field[:, :, 0]).clip(0, w - 1)
+    defMGy = (MGy + vec_field[:, :, 1]).clip(0, h - 1)
+
+    new_image = np.empty_like(image)
+
+    for channel in range(c):
+        # Get a linear interpolation for this color channel.
+        interpolation = RectBivariateSpline(hrange, wrange, image[channel],
+                                            kx=1, ky=1)
+
+        # grid=False since the deformed grid is irregular
+        new_image[channel] = interpolation(defMGy, defMGx, grid=False)
+    if color_axis == 2:
+        return _re_transpose_image(new_image)
+    else:
+        return new_image
+
+
+def _create_vec_field(fval, gradf, d1x, d2x, color_axis, smooth=0):
+    """Calculate the deformation vector field
+
+    In:
+    fval: float
+    gradf: numpy.ndarray
+        of shape Cxhxw with C=3 or C=1 (color channels), h,w >= 1.
```

Comment:
Here you talk about color_axis as 3 or 1 but later use python indexing, i.e. 0 or 2. Could you align this?

Reply:
@wielandbrendel not sure which other location you are referring to, but for the shape here, only 1 and 3 make sense, so that's correct (@EvgeniaAR but spaces around the equal sign and after the comma are missing :P )

Reply:
Sorry, you are right, I misread that part (I thought 1, 3 refers to channel axis, not the number of channels).